src/kernel.c

```diff
@@ -0,0 +1,642 @@
+		th->flags &= ~TH_FLAG_INLIST;
+		__thlist_add_pri(&th_ready, th);
+		KTHREAD_SWITCH_ISR();
+	}
+
+	if (DEBUG_KERNEL) debugf("[kernel] sleeping %ld %s[%p]\n", ticks, th->name, th);
+	disable_interrupts();
+
+	// Start a timer for the specified delay.
+	timer_link_t timer;
+	start_timer(&timer, ticks, TF_ONE_SHOT, cb);
+
+	// Mark the thread as being in a list to avoid the scheduler to panic.
+	// This is not "technically true", but it's "semantically true": when
+	// we say "a thread is in a list" it means that it's registered somewhere
+	// so that there's a way to wake it up in the future. In this case,
+	// the timer callback will do it, so we can reassure the kernel scheduler
+	// that everything is fine.
+	th->flags |= TH_FLAG_INLIST;
+
+	// Context switch
+	KTHREAD_SWITCH();
+
+	enable_interrupts();
+}
+
+void kthread_wait_event(kevent_t *evt)
+{
+	// Simplified API to wait for a single event. Just allocate a mailbox
+	// on the stack, attach to the event, and receive from it.
+	kmbox_t *mbox = kmbox_new_stack(1);
+	kmbox_attach_event(mbox, evt);
+	kmbox_recv(mbox);
+	kmbox_detach_event(mbox, evt);
+}
+
+kmbox_t* kmbox_new(int max_mails)
+{
+	// Allocate memory for the mailbox structure plus the mail array
+	int sz = sizeof(kmbox_t) + (max_mails)*sizeof(void*);
+	kmbox_t *mb = malloc(sz);
+	assertf(mb, "out of free memory");
+	memset(mb, 0, sz);
+	mb->max = max_mails;
+	return mb;
+}
+
+void kmbox_free(kmbox_t *mbox)
+{
+	assertf(mbox->evts == 0, "mbox_free() called, but missing %d mbox_detach_event() calls", mbox->evts);
+	free(mbox);
+}
+
+bool kmbox_empty(kmbox_t *mbox)
+{
+	return mbox->mails[mbox->r] == NULL;
+}
+
+bool kmbox_full(kmbox_t *mbox)
+{
+	return mbox->mails[mbox->w] != NULL;
+}
+
+/**
+ * @brief  Enqueue a mail into a mailbox (lower-level primitive).
+ *
+ * @note   This function must be called with interrupts disabled, as it
+ *         manipulates the ready list.
+ *
+ * @return 0 if the mail was not sent (mbox full)
+ *         1 if the mail was sent
+ *         2 if mail was sent and a higher-priority thread was awaken,
+ *         so a context switch should be done as soon as possible.
+ */
+int __kmbox_try_send(kmbox_t *mbox, void *mail)
+{
+	// If the mbox is full, there's nothing to do
+	if (kmbox_full(mbox))
+		return 0;
+
+	// Enqueue the new mail
+	mbox->mails[mbox->w++] = mail;
+	if (mbox->w == mbox->max)
+		mbox->w = 0;
+
+	// Awake pending threads. If a higher-priority thread is awaken,
+	// notify it to the caller.
+	return __thlist_splice_pri(&th_ready, &mbox->wait_list) ? 2 : 1;
+}
+
+bool kmbox_try_send(kmbox_t *mbox, void *mail)
+{
+	int ret;
+
+	disable_interrupts();
+	ret = __kmbox_try_send(mbox, mail);
+	if (ret == 2)
+		kthread_yield();
+	enable_interrupts();
+
+	return ret != 0;
+}
+
+void kmbox_send(kmbox_t *mbox, void *mail)
+{
+	disable_interrupts();
+	while (!kmbox_try_send(mbox, mail))
+	{
+		__thlist_add(&mbox->wait_list, th_cur);
+		KTHREAD_SWITCH();
```

Comment:
Yes. If the highest priority thread tries to write to a full mbox, it would keep spinning there. The only way to get out is that an interrupt fires and wakes up a thread that was waiting on the interrupt (kevent) and that thread is even higher priority so it gets scheduled, and it pulls from the mbox.

We might in theory detect this looping scenario, but I'm not sure what we would do in that case. Even if the above description sounds far fetched, it's not *that* weird. Think a main thread enqueuing RSP jobs for a RSP thread. The main thread pushes many jobs and at some point fills the mbox. The RSP threads pop them one at a time, schedule them in the RSP, and then wait execution (wait for RSP interrupt). I'm not sure there is nothing inherently wrong with this. And if the mbox is full, the main thread hasn't really anything else better to do than (spin)waiting.